File: Unit2-10/Java/Truck.java
/*
* This class creates a truck using methods from vehicles
*
* @author  Evan Beaudoin
* @version 1.0
* @since   2024-05-23
*/

/**
 * This is the class.
*/
public class Truck extends AbstractVehicle {
    /**
     * Variable for the class.
     *
     * @param licensePlateNumber the license plate number on the vehicle
    */
    private String licensePlateNumber = "";

    /**
     * Creates the base truck.
     *
     * @param startColour the base colour
     * @param startPlateNumber the base license plate number
     * @param startMaxSpeed the maximum speed of the vehicle
    */
    public Truck(
            String startColour, double startMaxSpeed, String startPlateNumber
    ) {
        super(startColour, startMaxSpeed, 1 + 1 + 1 + 1);
        this.licensePlateNumber = startPlateNumber;
    }

    /**
     * Gets the license plate number of the vehicle.
     *
     * @return the current licensePlateNumber
    */
    public String getLicensePlateNumber() {
        return this.licensePlateNumber;
    }

    /**
     * Creates a string about the status of the vehicle.
     *
     * @return the string about the vehicle's status
    */
    public String getStatus() {
        final String lineBreak = "\n";
        String status = " -> Speed: " + this.getSpeed() + lineBreak;
        status += " -> MaxSpeed: " + this.getMaxSpeed() + lineBreak;
        status += " -> Color: " + this.getColour() + lineBreak;
        status += " -> License Plate: " + this.licensePlateNumber + lineBreak;
        status += " -> # of Wheels: " + this.getWheels() + lineBreak;

        return status;
    }

    /**
     * Sets the license plate of the vehicle.
     *
     * @param newPlateNumber of the vehicle.
    */
    public void setLicensePlateNumber(String newPlateNumber) {
        this.licensePlateNumber = newPlateNumber;
    }

    /**
     * Brakes the truck, changing speed.
     *
     * @param brakePower of the brake
     * @param brakeTime of the brake
     * @param airPressure applied to the truck
    */
    public void brake(
            double brakePower, double brakeTime, double airPressure
    ) {
        this.setSpeed(
                this.getSpeed() - (
                    brakePower * brakeTime
                ) - (airPressure * brakeTime)
        );
        if (this.getSpeed() < 0) {
            this.setSpeed(0);
        }
    }
}

File: Unit2-10/Typescript/Truck.ts
/**
*
* This class creates a truck using methods from vehicles
*
* By:      Evan Beaudoin
* Version: 1.2
* Since:   2024-04-05
*/

import { Vehicle } from './Vehicle'

export class Truck extends Vehicle {
// Class variables
private _plateNumber: string

// Creates the base truck
public constructor(startColour: string, startMaxSpeed: number, startPlateNumber: string) {
    super(startColour, startMaxSpeed, 4, 0)
    this._plateNumber = startPlateNumber
}

// Sets a new plate number to the truck
public set plateNumber(newPlateNumber: string) {
    this._plateNumber = newPlateNumber
}

// Gets the current plate number
public get plateNumber(): string {
    return this._plateNumber
}

// Creates a string about the status of the truck
public get status(): string {
    let statusString = ` -> Speed: ${this.speed}\n`
    statusString += ` -> MaxSpeed: ${this.maxSpeed}\n`
    statusString += ` -> Color: ${this.colour}\n`
    statusString += ` -> License Plate: ${this.plateNumber}\n`
    statusString += ` -> # of Wheels: ${this.wheels}`

    return statusString
}

// Accelerates the truck, changing the speed
public accelerate(accelPower: number, accelTime: number): void {
    this._speed = (accelPower * accelTime) + this._speed
    if (this._speed > this._maxSpeed) {
    this._speed = this._maxSpeed
    }
}

// Brakes the truck, changing speed
public brake(brakePower: number, brakeTime: number, airPressure: number): void {
    this._speed = this._speed - (brakePower * brakeTime) - (airPressure * brakeTime)
    if (this._speed < 0) {
    this._speed = 0
    }
}
}